### PR TITLE
Window resize fix

### PR DESCRIPTION
**Pull request by jojolepro**
For review.
Known bug: Crashes when running a debug build and resizing(without --release).
This is due to gfx having a debug_assert in his code to resize the code.
https://github.com/gfx-rs/gfx/blob/master/src/core/src/pso.rs#L279

### CODE REVIEW

**Comment by Xaeroxe** [comment]
I am thoroughly confused.  That assert you link from gfx seems to have the express intent of panicking if the window size is changed at all.

File: amethyst_renderer/examples/sphere.rs [hunk 1]
```diff
@@ -62,7 +62,6 @@ fn main() {
         events.poll_events(|e| match e {
             Event::WindowEvent { event, .. } => {
                 match event {
-                    WindowEvent::KeyboardInput { .. } |
```
Comment (Xaeroxe) [inline amethyst_renderer/examples/sphere.rs]: Yeah this got annoying pretty quick.  Thanks!

**Comment by jojolepro** [comment]
I have no idea why they made that. "If you are debugging, and changing the dimension in the method to change the dimension, then crash"

File: amethyst_renderer/src/pipe/stage.rs [hunk 2]
```diff
@@ -38,16 +38,29 @@ pub struct Stage {
     clear_color: Option<[f32; 4]>,
     clear_depth: Option<f32>,
     enabled: bool,
-    passes: Vec<CompiledPass>,
+	/// Draw passes of the draw stage
+    pub passes: Vec<CompiledPass>,
```
Comment (Xaeroxe) [inline amethyst_renderer/src/pipe/stage.rs]: The indentation on this comment looks funny, can you fix that?

**Comment by jojolepro** [comment]
It works perfectly fine when running in release though, so I don't get why they are using that.

File: amethyst_renderer/src/lib.rs [hunk 3]
```diff
@@ -209,7 +220,37 @@ impl Renderer {
     pub fn create_pipe(&mut self, pb: PipelineBuilder) -> Result<Pipeline> {
         pb.build(&mut self.factory, &self.main_target)
     }
-
+	/// Updates the rendering target (useful for resizes)
+	/// Works only on tux for now >_<
```
Comment (Xaeroxe) [inline amethyst_renderer/src/lib.rs]: What would need to be done to get this working elsewhere?

File: amethyst_renderer/src/pipe/stage.rs (hunk 2 above)
Comment (jojolepro) [reply amethyst_renderer/src/pipe/stage.rs]: For some reason the comment is aligned on my side :/

Comment (Xaeroxe) [reply amethyst_renderer/src/pipe/stage.rs]: Make sure to use spaces, we don't use tab characters in this project.

**Comment by Xaeroxe** [comment]
https://gitter.im/gfx-rs/gfx?at=59b02ffcc101bc4e3a985642 this is the response I got when I asked about that line on gfx.  I'm not really sure what it means because I don't have much experience with gfx but you might be able to make sense of it.

File: amethyst_renderer/src/lib.rs [hunk 4]
```diff
@@ -157,7 +157,18 @@ impl Renderer {
     pub fn new(el: &EventsLoop) -> Result<Renderer> {
         RendererBuilder::new(el).build()
     }
-
+	///Get immutable main target
+	//pub fn get_main_target(&self)->Target{
+	//	return *self.main_target.clone();
+	//}
```
Comment (Xaeroxe) [inline amethyst_renderer/src/lib.rs]: Any reason why this is commented out?

File: amethyst_renderer/src/lib.rs (hunk 3 above)
Comment (jojolepro) [reply amethyst_renderer/src/lib.rs]: I actually forgot about that comment, I'm installing the windows sdk to test that now.

File: amethyst_renderer/src/lib.rs [hunk 5]
```diff
@@ -209,7 +220,37 @@ impl Renderer {
...
     }
-
+	/// Updates the rendering target (useful for resizes)
+	/// Works only on tux for now >_<
+	pub fn regen_target(&mut self)->Option<Target>{
+    	use gfx_window_glutin;
+    	let (new_color, new_depth) = gfx_window_glutin::new_views(&self.window);
+		if let Some(window_size) = self.window_size(){
+			let target = Target::new(
+    		    ColorBuffer {
```
Comment (Xaeroxe) [inline amethyst_renderer/src/lib.rs]: Alignment is off here too.

Comment (jojolepro) [reply amethyst_renderer/src/lib.rs]: I used tabs everywhere...

Comment (Xaeroxe) [reply amethyst_renderer/src/lib.rs]: rustfmt is your friend.

Comment (jojolepro) [reply amethyst_renderer/src/lib.rs]: Thanks

**Comment by Xaeroxe** [comment]
According to gfx that's not the "right" solution as apparently that assert is there to prevent the PixelTargetSet from changing size.  Meaning there's probably some code that assumes it won't change size.  The solution @kvark recommended was to recreate the PixelTargetSet.

**Comment by omni-viral** [comment]
@Xaeroxe yeah. Since `PixelTargetSet` is from _gfx_core_ it may be unsafe to tweak those values by hand.
It is created in _gfx_ thus it must be recreated by _gfx_. But I have no idea how to instruct _gfx_ to do so.

**Comment by jojolepro** [comment]
I have no idea how to recreate the PixelTarget either ;(

**Comment by jojolepro** [comment]
By the way I tested on windows, and when resizing, it created a stack overflow.

**Comment by jojolepro** [comment]
@omni-viral Adding out.pixel_targets = PixelTargetSet::new() in bake_to of PipelineData failed to avoid crashing.

**Comment by omni-viral** [comment]
@jojolepro do you have the code in your branch?

**Comment by jojolepro** [comment]
@omni-viral Pushed to the branch

File: amethyst_renderer/src/lib.rs (hunk 3 above)
Comment (jojolepro) [reply amethyst_renderer/src/lib.rs]: Fixed

**Comment by jojolepro** [comment]
examples/04_pong/main.rs
src/audio/play.rs

I didn't edit those files on my side, they can be discarted.

File: src/ecs/rendering/systems.rs [hunk 6]
```diff
@@ -54,14 +54,13 @@ impl<'a> System<'a> for RenderSystem {
                                 stage.set_target(new_target.clone());
                                 // Update stage effects, which are used to draw on the buffers
                                 for pass in stage.passes.iter_mut() {
-                                    let mut effect = &mut pass.effect.data;
+                                    let effect = &mut pass.effect.data;
                                     // Update Color Buffer
-                                    effect.out_colors.extend(
+                                    effect.out_colors =
```
Comment (omni-viral) [inline src/ecs/rendering/systems.rs]: `clear` + `extend` will reuse allocation. No big deal of course but why not.

Comment (jojolepro) [reply src/ecs/rendering/systems.rs]: I search "rust vec remove all" and didn't find anything, thanks for saving me :D

**Comment by Xaeroxe** [comment]
I'm not sure why that merge was necessary, but there we go.

**Comment by jojolepro** [comment]
Probably because I ran rustfmt just before I git add *

**Comment by Xaeroxe** [comment]
@omni-viral r+ this if you think it's good.

**Comment by jojolepro** [comment]
Doesn't work on Rhuagh's mac for some reason. Not sure if it could be related to the new resize code, or if its a general/separate engine bug for drawing on mac (he said he is not using metal features).
Any idea @omni-viral ?

**Comment by jojolepro** [comment]
Travis build because of a lag spike?
https://travis-ci.org/amethyst/amethyst/jobs/274783317

**Comment by omni-viral** [comment]
@jojolepro I'll test it myself today on mac

File: src/ecs/rendering/systems.rs [hunk 7]
```diff
@@ -38,6 +38,46 @@ impl<'a> System<'a> for RenderSystem {
...
         use std::time::Duration;
 
+        // FIXME issue when running in debug mode:
+        //https://github.com/gfx-rs/gfx/blob/master/src/core/src/pso.rs#L279
+        //Fetch current window size for this rendersystem
+        //Compare with current target size
+        if let Some(cur_window_size) = self.renderer.window_size() {
+            //Window size changed
+            for (name, target) in self.pipe.targets.iter_mut() {
+                if cur_window_size != target.size() {
```
Comment (omni-viral) [inline src/ecs/rendering/systems.rs]: `regen_target` creates target bound to window (the one you see).
Main target has a name "" (empty str). Thus only targets named "" must be replaced with new target bound to window. Other targets are created by `TargetBuilder` with `Factory`.
Their size should be changed only if they don't have custom size. See `TargetBuilder`'s `custom_size` field.

File: src/ecs/rendering/systems.rs [hunk 8]
```diff
@@ -38,6 +38,46 @@ impl<'a> System<'a> for RenderSystem {
...
+                                    // Update Depth Stencil
+                                    effect.out_depth = new_target.depth_buf().map(|db| {
+                                        (db.as_output.clone(), (0, 0))
+                                    });
+                                }
+                            }
+                        }
+                        if *name == "" {
+                            //Means this is the main target reference of the pipeline
+                            self.renderer.set_main_target(target.clone());
```
Comment (omni-viral) [inline src/ecs/rendering/systems.rs]: Here you place old target into `renderer.main_target` field. Were you intend to do that?

File: amethyst_renderer/src/lib.rs [hunk 9]
```diff
@@ -209,7 +216,43 @@ impl Renderer {
     pub fn create_pipe(&mut self, pb: PipelineBuilder) -> Result<Pipeline> {
         pb.build(&mut self.factory, &self.main_target)
     }
-
+	/// Updates the rendering target (useful for resizes)
+	/// Works only on tux for now >_<
+	pub fn regen_target(&mut self)->Option<Target>{
```
Comment (omni-viral) [inline amethyst_renderer/src/lib.rs]: This function isn't working with backends other than openGL.
For example `gfx_window_dxgi` has no function named `new_views`

Comment (jojolepro) [reply amethyst_renderer/src/lib.rs]: I'd be curious to know how that code even compiled to be honest.

**Comment by jojolepro** [comment]
@omni-viral @Xaeroxe https://gitter.im/gfx-rs/gfx?at=59b9302c1081499f1f281bec

**Comment by Xaeroxe** [comment]
@jojolepro Curious.  Do you think this could explain the behavior on @Rhuagh's machine?

**Comment by Rhuagh** [comment]
No, it's not. I'm not using feature metal

**Comment by jojolepro** [comment]
No, but that means that we can't have resizing for metal for now...

**Comment by Rhuagh** [comment]
Yes, but we also don't have much support for non-gl backend yet so that's fine.

File: amethyst_renderer/src/pipe/mod.rs [hunk 10]
```diff
@@ -98,11 +100,18 @@ impl Pipeline {
     pub fn enabled_stages(&self) -> Stages {
         Stages(self.stages.iter().filter(|s| s.is_enabled()))
     }
-
+	/// Get the mutable stages
+    pub fn stages_mut(&mut self)->&mut Vec<Stage>{
```
Comment (torkleyy) [inline amethyst_renderer/src/pipe/mod.rs]: Any reason why you made the field public **and** added a getter? Additionally, is every possible safe modification of the stages vector valid API usage?

File: amethyst_renderer/src/pipe/mod.rs [hunk 11]
```diff
@@ -98,11 +100,18 @@ impl Pipeline {
...
+        &mut self.stages
+    }
     /// Returns an immutable reference to all targets and their name strings.
     pub fn targets(&self) -> &HashMap<String, Target> {
         &self.targets
     }
+	/// Returns a mutable reference to all targets and their name strings.
+	pub fn targets_mut(&mut self) -> &mut HashMap<String,Target>{
```
Comment (torkleyy) [inline amethyst_renderer/src/pipe/mod.rs]: Same here

File: amethyst_renderer/src/pipe/pass.rs [hunk 12]
```diff
@@ -41,7 +41,7 @@ impl Debug for Description {
 
 #[derive(Clone)]
 pub struct CompiledPass {
-    effect: Effect,
+    pub effect: Effect,
```
Comment (torkleyy) [inline amethyst_renderer/src/pipe/pass.rs]: doc: made field public without adding docs; I'm not sure yet if it is necessary to expose this.

File: amethyst_renderer/src/pipe/stage.rs [hunk 13]
```diff
@@ -38,16 +38,29 @@ pub struct Stage {
...
     pub fn with_target<N: Into<String>>(target_name: N) -> StageBuilder {
         StageBuilder::new(target_name.into())
     }
-
+	/// Updates the stage target to a copy of some pipe.target
+	pub fn update_target(&mut self,updated:Target){
+		self.target = updated;
+	}
+	/// Get the target name, used to match them with pipe.targets
+	pub fn get_target_name(&self)->String{
```
Comment (torkleyy) [inline amethyst_renderer/src/pipe/stage.rs]: Getters in Rust do not have a `get_` prefix.

File: amethyst_renderer/src/pipe/stage.rs [hunk 14]
```diff
@@ -38,16 +38,29 @@ pub struct Stage {
...
 }
 
 impl Stage {
     /// Builds a new `Stage` which outputs to the `Target` with the given name.
     pub fn with_target<N: Into<String>>(target_name: N) -> StageBuilder {
         StageBuilder::new(target_name.into())
     }
-
+	/// Updates the stage target to a copy of some pipe.target
+	pub fn update_target(&mut self,updated:Target){
```
Comment (torkleyy) [inline amethyst_renderer/src/pipe/stage.rs]: Could we just make this a setter or is there anything else that could qualify this as an "update" in the future as opposed to a simple write?

**EDIT:** How does this differ from `set_target`?

File: amethyst_renderer/src/pipe/stage.rs [hunk 15]
```diff
@@ -160,9 +173,9 @@ impl StageBuilder {
 
     /// Builds and returns the stage.
     pub(crate) fn build(mut self, fac: &mut Factory, targets: &Targets) -> Result<Stage> {
-        let out = targets.get(&self.target_name).cloned().ok_or(
+        let out = targets.get(&self.target_name.clone()).cloned().ok_or(
```
Comment (torkleyy) [inline amethyst_renderer/src/pipe/stage.rs]: Any reason you clone a string which you create a reference to?

File: src/ecs/rendering/resources.rs [hunk 16]
```diff
@@ -4,7 +4,7 @@ use std::sync::Arc;
 
 use crossbeam::sync::MsQueue;
 use futures::{Async, Future, Poll};
-use futures::sync::oneshot::{Receiver, Sender, channel};
+use futures::sync::oneshot::{channel, Receiver, Sender};
```
Comment (torkleyy) [inline src/ecs/rendering/resources.rs]: Why has this been reordered?

File: src/ecs/rendering/systems.rs [hunk 17]
```diff
@@ -38,6 +38,46 @@ impl<'a> System<'a> for RenderSystem {
...
+                                        .map(|cb| cb.as_output.clone())
+                                    );
+                                    // Update Depth Stencil
+                                    effect.out_depth = new_target.depth_buf().map(|db| {
+                                        (db.as_output.clone(), (0, 0))
+                                    });
+                                }
+                            }
+                        }
+                        if *name == "" {
```
Comment (torkleyy) [inline src/ecs/rendering/systems.rs]: nit: should be `name.is_empty()`

**Comment by torkleyy** [comment]
I'm requesting a review from either @ebkalderon or @kvark before merging this. I do not know what is the recommended strategy for resizing and I cannot validate the current resizing logic, also considering I don't have enough for further investigation right now (as I said on Gitter lately, won't be too active these weeks unfortunately). I don't think we should choose a solution which limits our support for backends.

File: src/ecs/rendering/resources.rs (hunk 16 above)
Comment (jojolepro) [reply src/ecs/rendering/resources.rs]: rustfmt

Comment (torkleyy) [reply src/ecs/rendering/resources.rs]: I think we agreed on having `Type`s before `function`s.

cc @Rhuagh

**Comment by derekdreery** [comment]
If there's no impact on the API, we can always replace the solution later. Could just disable the code on unsupported backends for now? I haven't read all the code - there may be reasons why this is not possible.

File: amethyst_renderer/src/lib.rs (hunk 9 above)
Comment (omni-viral) [reply amethyst_renderer/src/lib.rs]: You aways tried it with glutin

**Comment by omni-viral** [comment]
Disable resizing where it is unsupported (metal).
Fix issues on other backends on all platforms where possible.

File: amethyst_renderer/src/pipe/pass.rs (hunk 12 above)
Comment (Xaeroxe) [reply amethyst_renderer/src/pipe/pass.rs]: We have the deny_documentation lint turned on so if this is exposed then code won't compile.

File: src/ecs/rendering/resources.rs (hunk 16 above)
Comment (Xaeroxe) [reply src/ecs/rendering/resources.rs]: Honestly I'm so over burning development time on how to sort <5 item import lists.

**Comment by jojolepro** [comment]
Since basically the whole pipeline was rewritten, that code is going to go in the trash, except if the pipeline changes will take a long time before release. If that's the case, this could get released as a bug fix until the new pipeline is released.

**Comment by omni-viral** [comment]
I plan to finish #317 in a couple of days

**Comment by Xaeroxe** [comment]
@jojolepro I'm closing this, please re-open it if you feel I was in error.

**Comment by jojolepro** [comment]
Rest in peperoni

**Pull request closed without merging**